protocol/network/4-architecture.md

```diff
@@ -0,0 +1,126 @@
+4 Architecture
+==============
+
+libp2p was designed around the Unix Philosophy, creating smaller components, easier to understand and to test. These components should also be able to be swapped in order to accomodate different technologies or scenarios and also make it that it is upgradable over time.
+
+Although different Peers can support different protocols depending on their capabilities, any Peer can act as a dialer and/or a listener for connections from other Peers, connections that once established can be reused from both ends, removing the distinction between clients and servers.
+
+libp2p interface acts as a thin veneer to three subsystems that are required in order for peers to be able to communicate. These subsystems are allowed to be built on top of other subsystems as long as they respect the standardized interface. The main 3 subsystems are:
+
+- Peer Routing - Mechanism to find a Peer in a network. This Routing can be done recursively, iteratively or even in a broadcast/multicast mode.
+- Swarm - Handles everything that touches the 'opening a stream' part of libp2p, from protocol muxing, stream muxing, NAT Traversal, Connection Relaying, while being multitransport
+- Distributed Record Store - A system to store provider records and relay records
+
+Each of these subsystem exposes a well known interface (see chapter 6 for Interfaces) and may use eachother in order to fulfil their goal. A global overview of the system is:
+
+```
+┌─────────────────────────────────────────────────────────────────────────────────┐
+│                                  libp2p                                         │
+└─────────────────────────────────────────────────────────────────────────────────┘
+┌─────────────────────────┐┌──────────────────────────┐┌──────────────────────────┐
+│       Peer Routing      ││      Swarm               ││ Distributed Record Store │
+└─────────────────────────┘└──────────────────────────┘└──────────────────────────┘
+```
```

Comment:
> the "not three systems" above changes this picture to have other things (namely, NAT, and Relay). there are likely to be other protocols "that peer" here too. and the reason "NAT" and "Relay" can't be part of "the thing named Swarm here" is that "NAT" and "Relay" make use of "Peer Routing" and "Peer Routing" makes use of "NAT" and "Relay". They're really peering-protocols.

Oh, I believe I'm starting to see what you mean. NAT and Relay can use Peer Routing because of the `contract`that any of these subsystems will respect the same interface and expectations. For e.g Kad-Record-Store uses Kad-Routing to perform it's Kademlia queries 

I actually thought this was pretty neat and was pretty happy when I understood I could break logic apart by defining expectations. libp2p becomes a big composition of other modules